Smith-Waterman Algorithm

ragnorc opened the pull request.

TimothyStiles commented on align/align.go:
```diff
@@ -91,6 +82,74 @@ func NeedlemanWunsch(stringA string, stringB string, scoring Scoring) (int, stri
+			alignB = string(stringB[j-1]) + alignB
+			i--
+			j--
+		} else if matrix[i][j] == matrix[i-1][j]+scoring.GapPenalty {
```
this condition isn't covered by tests

ragnorc replied:
Good point. Just changed the test case to have gaps on both alignment strings: [98b16bf](https://github.com/TimothyStiles/poly/pull/292/commits/98b16bf216a97262726d7427884661a02c7cdf11)

TimothyStiles commented on align/align.go:
```diff
@@ -91,6 +82,74 @@ func NeedlemanWunsch(stringA string, stringB string, scoring Scoring) (int, stri
+			alignA = "-" + alignA
+			alignB = string(stringB[j-1]) + alignB
+			j--
+		} else {
```
is there any case where this can happen? Shouldn't panic but should return an error if this case is possible.

ragnorc replied:
I am pretty sure this case should never happen. It would mean that the matrix was constructed erroneously or there is a bug somewhere.

TimothyStiles commented:
@ragnorc this was super helpful! I've added some documentation and formatting and refactored some variables. Going to merge this and then we can work on adding weight matrices mentioned in #290.

ragnorc commented:
> Thank you @ragnorc this PR was excellent! I've added some documentation and formatting and refactored some variables. Going to merge this and then we can work on adding weight matrices mentioned in #290.

Awesome, thank you! Working on the weight matrices now.
